### DISCUSSION TITLE
Expo ImagePicker - Freeform image rotation and crop on iOS and Android

### DISCUSSION THREAD

**Discussion by stharvey** [expo/expo#26849] [Proposals and Ideas]
Are there any solutions, or plans to implement in expo-image-picker the function for freeform image rotation and cropping?

The 90 degree rotation and square cropping in iOS is quite limiting.

Users often take photos and would like to make small adjustments to rotation to get a horizontal horizon but I haven't found a way to implement this in Expo at the moment.

**Comment by tsapeta** [comment]
Have you tried `expo-image-manipulator`?

**Reply by stharvey** [reply]
Thanks for the suggestion.  I already use `expo-image-manipulator` for image resizing and compression but it's only API, there's no UI.  I'm looking for something will allow users to rotate and crop quickly.  I don't think there's anything that exists so I've started work on creating my own UI for `expo-image-manipulator`

Thanks again.

**Comment by stharvey** [comment]
I've now created a UI for doing this using `expo-image-manipulator` - at some point I'll make the code available.

![crop rotate](https://github.com/expo/expo/assets/23450702/cb20e078-dda0-49f6-9f24-b432ce7d455c)

**Reply by Code-Victor** [reply]
Please, do. I'm currently facing the same problem and it is a pain to deal with.
A Github gist can help point in the right direction.

**Reply by VictorRutskin** [reply]
@stharvey sharing this nice ui would be lovely!

**Reply by Manavjain25** [reply]
@stharvey can u pls share the ui, thanks !!

**Comment by mateofowlerz** [comment]
I know is not the same, but u can use: https://www.npmjs.com/package/react-native-image-crop-picker for free crop. Is slower btw

**Reply by Manavjain25** [reply]
we need for expo managed apps not native

**Comment by IggsGrey** [comment]
Not sure if I have discovered a 'bug' or an undocumented feature.

Apparently passing null to the `aspect` param gives you a freeform crop.

```
let result = await ImagePicker.launchImageLibraryAsync({
	mediaTypes: ['images'],
	allowsEditing: true,
	aspect: null,
	quality: 1,
});
```

This works for `launchCameraAsync` as well.

I'm yet to review the expo-image-picker code to find out why this works.

I discovered this while running within expo go on android. Not yet tested on an ios device.

**Reply by Manavjain25** [reply]
u can't pass null to aspect right it expects [number,number] | undefined. How did u do that ??

**Reply by IggsGrey** [reply]
Not sure what you mean by how did I do that.. but basically typing `null` instead of passing a number array or undefined does not break it. It just gave me a free-form crop. I'm not using this in production, though.

So again, not sure if it's an undocumented feature, cos like you said it should only accept array or undefined.

**Reply by Manavjain25** [reply]
Did u test this on ios ??, because on android if u don't pass aspect then also it will give u freeform crop instead of passing null.

**Reply by Brian4804** [reply]
You can use undefined: 
  const res = await ImagePicker.launchCameraAsync({
    mediaTypes: (ImagePicker as any).MediaType?.Image ?? ImagePicker.MediaTypeOptions.Images,
    allowsEditing: true,
    quality: 1,
    aspect: undefined,
  });
  if (res.canceled) return;

**Reply by pboos** [reply]
Note that this has no effect on iOS as it will always use square crop as per documentation:
> This is only applicable on Android, since on iOS the crop rectangle is always a square.
